File: backend/app/schemas/rating.py
"""
Rating schemas
"""
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime


class SubmitRatingRequest(BaseModel):
    """Request schema for submitting ride rating"""
    booking_id: str = Field(..., description="Booking ID")
    rating: int = Field(..., ge=1, le=5, description="Rating from 1 to 5 stars")
    comment: Optional[str] = Field(default="", max_length=500, description="Optional comment")


class RatingResponse(BaseModel):
    """Response schema for rating"""
    id: str
    booking_id: str
    user_id: str
    driver_id: str
    rating: int
    comment: str
    created_at: datetime

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class DriverRatingStatsResponse(BaseModel):
    """Response schema for driver rating statistics"""
    driver_id: str
    average_rating: float
    total_ratings: int
    rating_breakdown: dict  # {"5": 10, "4": 5, "3": 2, "2": 1, "1": 0}

File: mobile/services/rating.ts
/**
 * Rating API service
 */
import api from './api';

export interface Rating {
  id: string;
  booking_id: string;
  user_id: string;
  driver_id: string;
  rating: number;
  comment: string;
  created_at: string;
}

export interface SubmitRatingRequest {
  booking_id: string;
  rating: number;
  comment?: string;
}

export interface DriverRatingStats {
  driver_id: string;
  average_rating: number;
  total_ratings: number;
  rating_breakdown: {
    '5': number;
    '4': number;
    '3': number;
    '2': number;
    '1': number;
  };
}

/**
 * Submit a rating for a completed ride
 */
export async function submitRating(data: SubmitRatingRequest): Promise<Rating> {
  const response = await api.post<Rating>('/ratings/submit', data);
  return response.data;
}

/**
 * Get rating statistics for a driver
 */
export async function getDriverRatingStats(driverId: string): Promise<DriverRatingStats> {
  const response = await api.get<DriverRatingStats>(`/ratings/driver/${driverId}/stats`);
  return response.data;
}

/**
 * Get rating for a specific booking
 */
export async function getBookingRating(bookingId: string): Promise<Rating> {
  const response = await api.get<Rating>(`/ratings/booking/${bookingId}`);
  return response.data;
}

/**
 * Get all ratings submitted by current user
 */
export async function getMyRatings(limit: number = 20, skip: number = 0): Promise<Rating[]> {
  const response = await api.get<Rating[]>('/ratings/my-ratings', {
    params: { limit, skip }
  });
  return response.data;
}
